californium-core/src/main/java/org/eclipse/californium/core/CoapClient.java

```diff
@@ -846,7 +849,15 @@ private Request ifNoneMatch(Request request) {
 	private CoapObserveRelation observeAndWait(Request request, CoapHandler handler) {
 		Endpoint outEndpoint = getEffectiveEndpoint(request);
 		CoapObserveRelation relation = new CoapObserveRelation(request, outEndpoint);
-		request.addMessageObserver(new ObserveMessageObserverImpl(handler, relation));
+		// add message observer to get the response.
+		ObserveMessageObserverImpl messageObserver = new ObserveMessageObserverImpl(handler, relation);
+		request.addMessageObserver(messageObserver);
+		// add notification listener to all notification
+		NotificationListener notificationListener = new Adapter(messageObserver, request);
+		outEndpoint.addNotificationListener(notificationListener);
+		// messageobserver should remove this listener when the request is cancelled
+		messageObserver.setNotificationListener(notificationListener);
+		relation.setNotificationListener(notificationListener);
```

Comment:
You're right @jvermillard , but the high level API (`CoapClient`) is not really cluster friendly :/.

@sophokles73, I don't think this is a good idea to force ObservationStore implementation to keep in memory a reference to the original request with its `MessageObserver` . [Here](https://github.com/eclipse/leshan/blob/clustering/leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisObservationRegistry.java#L330) an example of the leshan redis implementation. This seems to me